### PR TITLE
Pathfinding Improvements: Improve 5-10-5 Diagonal Handling

### PR DESCRIPTION
**Pull request by JDCalvert**
The pathfinding algorithm currently handles 5-10-5 diagonals by storing each node twice, once for being a "5" diagonal and once for being a "10". This potentially doubles the number of iterations we need to make.

I've removed this and changed all diagonals to be worth 1.5 squares when using 5-10-5 diagonals. This makes two diagonals add to 3 squares. This also has the advantage that the algorithm will favour being in a position where the "next" diagonal only costs 1 square (e.g. current cost is 10, next will be 11.5, or 11 when rounded down) rather than 2 (e.g. current cost is 10.5, next will be 12). If the previous waypoints have a combined odd number of diagonals, then we change the cost of the starting node to 0.5 to simulate the same behaviour.

**Note: ** This currently includes the changes from #169 as it builds upon them. Once #169 is merged, this PR will only include the changes above.

### CODE REVIEW

**Comment by manuelVo** [comment]
I'm not sure about this. I'm planning to add support difficult terrain in pathfinding in the future. This will require tapping into Terrain Ruler's API, which will take start, end, and an info whether the next diagonal will cost 5 or 10 and will return the cost of the movement and the new info whether the next diagonal will cost 5 or 10. Currently, I don't have any ideas how that API could be integrated with the change proposed here.

**Comment by JDCalvert** [comment]
> I'm not sure about this. I'm planning to add support difficult terrain in pathfinding in the future. This will require tapping into Terrain Ruler's API, which will take start, end, and an info whether the next diagonal will cost 5 or 10 and will return the cost of the movement and the new info whether the next diagonal will cost 5 or 10. Currently, I don't have any ideas how that API could be integrated with the change proposed here.

With the new method, if the `cost` is `x.5` then the next diagonal will cost 10, and if it's `x.0` then the next diagonal will cost 5.

For example, if the current node's cost is 4.5, then we can call Terrain Ruler and tell it the next diagonal will cost 10. If Terrain Ruler returns with a cost of 3 and the next diagonal will be 5, then we can interpret that as a cost of 2.5 (3 movement, -0.5 because we went from the next diagonal costing 10 to costing 5) and the cost of the next space is 7.

The opposite is also true - if we start with a cost of 4 and next diagonal is a 5, and Terrain Ruler returns a cost of 3 and next diagonal is 10, then we can interpret that as a cost of 3.5 (3 movement, +0.5 because we went from the next diagonal being 5 to next diagonal being 10.

Or, Terrain Ruler can update to accept and return fractional

**Comment by manuelVo** [comment]
This might actually work. At the moment I cannot think of any edge case where the new logic would break. I'll give it a bit more thought, but your proposal sounds very reasonable and advantageous.

File: js/pathfinding.js [hunk 1]
```diff
@@ -149,8 +137,14 @@ function calcNoDiagonals(waypoints) {
...
 
+/**
+ * Estimate the travel distance between two points, as the crow flies. Most of the time, this is 1
+ * per space, but for a square grid using 5-10-5 diagonals, count each diagonal as an extra 0.5
+ */
 function estimateCost(pos, target) {
-	return Math.max(Math.abs(pos.x - target.x), Math.abs(pos.y - target.y));
+	const distX = Math.abs(pos.x - target.x);
+	const distY = Math.abs(pos.y - target.y);
+	return Math.max(distX, distY) + (use5105 ? Math.min(distX, distY) * 0.5 : 0);
```
Comment (manuelVo) [inline js/pathfinding.js:147]: This logic is flawed. For a single diagonal step on 5/10/5 this will estimate the cost to be 1.5, while the actual distance is 1. For A* to work correctly it's important that the estimated cost is always equal or cheaper to the actual cost, which isn't the case here. Options would be to either revert this change or modifying the new logic to ensure that it cannot ever overestimate the distance.

**Review by manuelVo — CHANGES_REQUESTED** [review]

File: js/pathfinding.js (hunk 1 above)
Comment (JDCalvert) [reply js/pathfinding.js:147]: The way I've changed the logic in `getNode` means that a single diagonal step actually does cost `1.5`, not either `1.0` or `2.0`. The difference is important for the algorithm - it is better to be in a space where the next diagonal will cost 5 ft. of movement (i.e. in the algorithm the cost is `x.0`) than to be in the same space where the next diagonal will cost 10 ft. (i.e. in the algorithm the cost is `x.5`), and that is reflected in the `.5` being part of both the cost and estimated cost.

Even at the end of the path, the difference _could_ be important (if you want to place a waypoint and carry on moving). Only when you're finished moving does the difference between `x.5` and `x.0` stop mattering.

Comment (manuelVo) [reply js/pathfinding.js:147]: Yes, you're right.
